### PR TITLE
[DatePicker] Update component to use HTML table structure

### LINKED ISSUE #797: [a11y] Datepicker structure not clearly conveyed
## Issue summary

The [Datepicker](https://polaris.shopify.com/components/forms/date-picker#navigation) structure is not conveyed clearly to screen reader users.

* VoiceOver skips the table caption and headers, probably because they don’t have roles associated with the grid, and doesn’t report column/row numbers correctly.
* VO/macOS says "clickable" and NVDA says "editable" for cells but doesn't convey the table structure.
* VO does not convey which cell is selected.
* `role="rowheader"` is used for the whole row like `<thead>`, not as individual headers (equivalent of `<th scope="row">`).
* The `aria-label` values are not read.

#### Datepicker output (HTML)

```
<div class="Polaris-DatePicker__MonthContainer">
  <div role="grid" class="Polaris-DatePicker__Month">
    <div class="Polaris-DatePicker__Title">February 2018</div>
    <div role="rowheader" class="Polaris-DatePicker__WeekHeadings">
      <div aria-label="Sunday" class="Polaris-DatePicker__Weekday">Su</div>
      <div aria-label="Monday" class="Polaris-DatePicker__Weekday">Mo</div>
      <div aria-label="Tuesday" class="Polaris-DatePicker__Weekday">Tu</div>
      <div aria-label="Wednesday" class="Polaris-DatePicker__Weekday">We</div>
      <div aria-label="Thursday" class="Polaris-DatePicker__Weekday">Th</div>
      <div aria-label="Friday" class="Polaris-DatePicker__Weekday">Fr</div>
      <div aria-label="Saturday" class="Polaris-DatePicker__Weekday">Sa</div>
    </div>
    <div role="row" class="Polaris-DatePicker__Week">
      <div class="Polaris-DatePicker__EmptyDay"></div>
      <div class="Polaris-DatePicker__EmptyDay"></div>
      <div class="Polaris-DatePicker__EmptyDay"></div>
      <div class="Polaris-DatePicker__EmptyDay"></div>
      <button type="button" tabindex="0" class="Polaris-DatePicker__Day" aria-label="February 1 2018" aria-selected="false" role="gridcell">1</button>
      <button type="button" tabindex="-1" class="Polaris-DatePicker__Day" aria-label="February 2 2018" aria-selected="false" role="gridcell">2</button>
      <button type="button" tabindex="-1" class="Polaris-DatePicker__Day" aria-label="February 3 2018" aria-selected="false" role="gridcell">3</button>
    </div>
    <div role="row" class="Polaris-DatePicker__Week">
      <button type="button" tabindex="0" class="Polaris-DatePicker__Day" aria-label="February 4 2018" aria-selected="false" role="gridcell">4</button>
      <button type="button" tabindex="-1" class="Polaris-DatePicker__Day" aria-label="February 5 2018" aria-selected="false" role="gridcell">5</button>
      <button type="button" tabindex="-1" class="Polaris-DatePicker__Day" aria-label="February 6 2018" aria-selected="false" role="gridcell">6</button>
      <!-- ... -->
    </div>
    <!-- ... -->
  </div>
</div>
```

## Steps to reproduce

1. Navigate to [Datepicker](https://polaris.shopify.com/components/forms/date-picker#navigation).
1. Using VoiceOver or NVDA, read through the table using reading keys or the arrow keys to navigate.
1. Compare to the example [Data Grid Examples](https://www.w3.org/TR/wai-aria-practices-1.1/examples/grid/dataGrids.html) from the ARIA authoring guidelines.

### Expected

- The table structure and controls will be clearly conveyed.

### Actual

- The table structure and controls are not clearly conveyed.

See [VoiceOver videos](https://drive.google.com/drive/folders/1Ar8Fj5ZpiIW2eElpapcZfOy2nhRHpmxY) for a comparison.

## Recommendations

* Use a standard HTML table for the datepicker structure, and enhance the `<table>` element with `role="grid"`. This will allow the rest of the table content function within the `grid` structure automatically.
* Use a separate `<button>` with `aria-pressed` to convey the selected state for each available date control.
* See the [ARIA spec](https://www.w3.org/TR/wai-aria-1.1/#grid) and the [ARIA authoring guidelines](https://www.w3.org/TR/wai-aria-practices-1.1/#grid) for `grid`.
* VoiceOver will read `aria-label` values assign to `<th>` elements, but NVDA does not appear to do the same. Looking into an alternative solution is recommended.

### Recommended table structure output (HTML)

```
<div class="Polaris-DatePicker__MonthContainer">
  <table role="grid" class="Polaris-DatePicker__Month">
    <caption class="Polaris-DatePicker__Title">February 2018</caption>
    <tr class="Polaris-DatePicker__WeekHeadings">
      <th scope="col" class="Polaris-DatePicker__Weekday">Su</th>
      <th scope="col" class="Polaris-DatePicker__Weekday">Mo</th>
      <th scope="col" class="Polaris-DatePicker__Weekday">Tu</th>
      <th scope="col" class="Polaris-DatePicker__Weekday">We</th>
      <th scope="col" class="Polaris-DatePicker__Weekday">Th</th>
      <th scope="col" class="Polaris-DatePicker__Weekday">Fr</th>
      <th scope="col" class="Polaris-DatePicker__Weekday">Sa</th>
    </tr>
    <tr class="Polaris-DatePicker__Week">
      <td class="Polaris-DatePicker__EmptyDay"></td>
      <td class="Polaris-DatePicker__EmptyDay"></td>
      <td class="Polaris-DatePicker__EmptyDay"></td>
      <td class="Polaris-DatePicker__EmptyDay"></td>
      <td>
        <button tabindex="0" class="Polaris-DatePicker__Day" aria-label="February 1 2018" aria-pressed="false">1</button>
      <td>
      <td>
        <button tabindex="-1" class="Polaris-DatePicker__Day" aria-label="February 2 2018" aria-pressed="true">2</button><!-- for selected dates -->
      </td>
      <td>
        <button tabindex="-1" class="Polaris-DatePicker__Day" aria-label="February 3 2018" aria-pressed="false">3</button>
      </td>
    </tr>
    <!-- ... -->
  </html>
</div>
```

## Specifications

- **Component:** Datepicker
- **WCAG SCs:** [1.3.1 Info and Relationships (Level A)](https://www.w3.org/TR/WCAG21/#info-and-relationships), [4.1.2 Name, Role, Value (Level A)](https://www.w3.org/TR/WCAG21/#name-role-value)
- **Testing done with:**
  - VoiceOver/Safari/iOS
  - Voiceover/Safari/macOS
  - NVDA/Firefox/Windows 10

### PR DESCRIPTION
**Pull request by dfmcphee**
### WHY are these changes introduced?

Fixes #797 

Currently the DatePicker component is misusing role attributes and the result is that it doesn't convey the proper structure to screen readers.

### WHAT is this pull request doing?

This refactors the DatePicker component to use HTML table markup instead of trying to use divs with role attributes.

## 

### How to 🎩

🖥 [Local development instructions](https://github.com/Shopify/polaris-react/blob/master/README.md#development)
🗒 [General tophatting guidelines](https://github.com/Shopify/polaris-react/blob/master/documentation/Tophatting.md)
📄 [Changelog guidelines](https://github.com/Shopify/polaris-react/blob/master/documentation/Versioning%20and%20changelog.md)

Look at the DatePicker examples in Storybook and make sure all still looks and works as expected.

</details>

### 🎩 checklist

* [x] Tested on [mobile](https://github.com/Shopify/polaris-react/blob/master/documentation/Tophatting.md#cross-browser-testing)
* [x] Tested on [multiple browsers](https://help.shopify.com/en/manual/intro-to-shopify/shopify-admin/supported-browsers)
* [x] Tested for [accessibility](https://github.com/Shopify/polaris-react/blob/master/documentation/Accessibility%20testing.md)

### CODE REVIEW

**Comment by dfmcphee** [comment]
One small thing I noticed is that the borders don't quite work the same way as before since previously we were using z-index and negative margins to pull the correct border over top, but now that we are using a table and `border-collapse: collapse` we aren't able to do that.

## Old structure
Screen Shot 2020-09-23 at 12 56 58 PM


## New structure
Screen Shot 2020-09-23 at 12 43 43 PM


In my opinion this is a fine trade-off for the accessibility improvements we get, especially since these styles will be going away and this isn't an issue in the new design language.

File: src/components/DatePicker/DatePicker.scss [hunk 1]
```diff
@@ -28,29 +25,41 @@ $in-range-border-color: #9ca6de;
   margin-left: spacing();
   max-width: calc(100% - #{spacing()});
   min-width: rem(230px);
+  table-layout: fixed;
```
Comment (alex-page) [inline src/components/DatePicker/DatePicker.scss:28]:
```suggestion
  width: 100%;
  table-layout: fixed;
```

I think you might need to stretch the container by default for FireFox.

Screen Shot 2020-09-23 at 1 33 28 PM

File: UNRELEASED.md [hunk 2]
```diff
@@ -9,6 +9,7 @@ Use [the changelog guidelines](https://git.io/polaris-changelog-guidelines) to f
 ### Enhancements
 
 - Hide `ActionMenu`, `Actions`, `RollupActions` menu popover overlays when printing ([#3277](https://github.com/Shopify/polaris-react/pull/3277))
+- Updated `DatePicker` component to use HTML table structure for better semantics ([#3303](https://github.com/Shopify/polaris-react/pull/3303))
```
Comment (alex-page) [inline UNRELEASED.md:12]:
```suggestion
- Updated `DatePicker` component to use semantic HTML table structure ([#3303](https://github.com/Shopify/polaris-react/pull/3303))
```

**Review by alex-page — APPROVED** [review]
Tophatted in Firefox, only found one issue. Thanks @dfmcphee

**Review by kyledurand — APPROVED** [review]
Tables! In 2020! I'm all for it 

🎩  , code, and markup all look great. I think we might want to follow up to this with a contrast improvement on selected date ranges though

image

**Comment by alex-page** [comment]
> follow up to this with a contrast improvement on selected date ranges though

I agree, but from what I can see in the admin it's usually on a white background.

**Pull request merged**